src/Ipac_Table_Writer/internals.cxx

```diff
@@ -214,48 +461,47 @@ void tablator::Ipac_Table_Writer::write_single_record_by_offset(
-
-            for (size_t element = 0; element < effective_array_size; ++element) {
-                os << IPAC_COLUMN_SEPARATOR << std::setw(width);
-                os << (null_value.empty() ? DEFAULT_NULL_VALUE : null_value);
-            }
-        } else if (active_datatype == tablator::Data_Type::UINT8_LE) {
-            // Do this case manually because write_type_as_ascii()
-            // isn't equipped to write bytes as ints, as IPAC_FORMAT
-            // requires.
-            size_t base_offset = curr_row_offset + table.offsets[i];
-            uint8_t const* curr_data = table.data.data() + base_offset;
-            size_t element_size = data_size(active_datatype);
-
-            for (size_t element = 0; element < column.array_size; ++element) {
-                os << IPAC_COLUMN_SEPARATOR << std::setw(width);
-                os << static_cast<const uint16_t>(*curr_data);
-                curr_data += element_size;
-            }
-        } else {
-            size_t base_offset = curr_row_offset + table.offsets[i];
-            uint8_t const* curr_data = table.data.data() + base_offset;
+    for (size_t col_id = 1; col_id < columns.size(); ++col_id) {
+        write_single_value(table, os, col_id, curr_row_offset, datatypes_for_writing);
+    }
+    os << IPAC_COLUMN_SEPARATOR << "\n";
+}
 
-            os << IPAC_COLUMN_SEPARATOR << std::setw(width);
-            std::stringstream ss_temp;
-            write_type_as_ascii(ss_temp, active_datatype, column.array_size, curr_data,
-                                width);
-            std::string s(ss_temp.str());
-            os << boost::replace_all_copy(s, "\n", " ");
+/**********************************************************/
+
+void tablator::Ipac_Table_Writer::write_single_record_by_offset(
+        const Table& table, std::ostream& os,
+        const std::vector<size_t>& included_column_ids, size_t curr_row_offset,
+        const std::vector<Data_Type>& datatypes_for_writing) {
+    size_t final_row_offset = table.data.size() - table.row_size();
+    if (curr_row_offset > final_row_offset) {
+        std::string msg = "invalid row_offset ";
+        msg.append(std::to_string(curr_row_offset))
+                .append("; largest possible is ")
+                .append(std::to_string(final_row_offset))
+                .append(".");
+        throw std::runtime_error(msg);
+    }
+
+    const std::vector<size_t> ipac_column_widths = get_column_widths(table);
+    const std::vector<Column>& columns = table.columns;
+
+    // Skip the null bitfield flag
+    std::vector<size_t>::const_iterator cols_iter = included_column_ids.begin();
+    if (cols_iter != included_column_ids.end() && *cols_iter == 0) {
+        ++cols_iter;
+    }
+    for (/* */; cols_iter != included_column_ids.end(); ++cols_iter) {
+        auto curr_col_id = *cols_iter;
+        if (curr_col_id > columns.size()) {
+            // shouldn't happen
+            break;
```

Comment:
Ah, right, I knew there would be fallout from switching from set to vector, but I forgot to follow up.  Do you have a preference for silently skipping invalid values vs. throwing exceptions?

Reply:
I have a slight preference for skipping invalid indexes, but I don’t feel strongly about it.

Reply:
On sober reflection, I'm fine with skipping invalid column indices, because the headers are there to tell the caller which columns actually appear in the response, but missing rows would be harder to detect and could mess up whatever further parsing the caller means to do.